Invalid WTF::URLs should not convert to NSURLs

achristensen07 opened the pull request:
#### 2404832a74ce484dc64a981d5d2ef659db3ccbef
<pre>
Invalid WTF::URLs should not convert to NSURLs
<a href="https://bugs.webkit.org/show_bug.cgi?id=286926">https://bugs.webkit.org/show_bug.cgi?id=286926</a>

Reviewed by John Wilander.

Creating valid NSURLs from invalid WTF::URLs has long been a source
of issues that we find more targeted fixes for.  We do this over and
over again in different places.  This will fix all of them.

A sprinkling of little fixes is needed to keep some of the failures
the same as they were before, but this also intends to change some
of the URLs accessed after putting them into WebKit when WebKit
thinks they are not valid.

To mitigate the binary compatibility effects of this change, I did
the important changes behind a linkedOnOrAfterSDKWithBehavior check.

* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
(WTF::URL::fileSystemPath const):
(WTF::URL::emptyCFURL): Deleted.

Canonical link: <a href="https://commits.webkit.org/289946@main">https://commits.webkit.org/289946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c64b99764c957c84cdfbf72fd27e86edf5b08949

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93451 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39247 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90544 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16196 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25969 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91495 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80046 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/48622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6202 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38355 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81292 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95293 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87269 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15668 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15924 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76383 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20772 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19193 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/8707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13834 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15684 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109762 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15425 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26392 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
```diff
@@ -346,7 +346,7 @@ static void configureRequestWithData(NSMutableURLRequest *request, const Resourc
 {
     if (isNull()) {
-        ASSERT(!m_nsRequest);
+        m_nsRequest = nil;
```
I assume this releases the request.

achristensen07 replied:
Yep.  There were a few cases where we have a URL that now converts to null from a non-null request.  In such cases, we want to have the request be null too.

johnwilander commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2041,6 +2041,18 @@ void WebPageProxy::loadRequestWithNavigationShared(Ref<WebProcessProxy>&& proces
 
     auto url = request.url();
+#if PLATFORM(COCOA)
+    if (!url.isNull() && !url.isValid() && WTF::linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ConvertsInvalidURLsToNull)) {
```
I prefer these kind of compound boolean clauses to be put in a convenience variable (or two) to improve readability. Especially when there are multiple negations in the clause like this one. Please consider something like `auto urlValidAndNotNull = !url.isNull() && !url.isValid();`.

johnwilander reviewed (approved):
r+ with one request for a readability change and one question on object deletion.

achristensen07 committed:
Invalid WTF::URLs should not convert to NSURLs
https://bugs.webkit.org/show_bug.cgi?id=286926

Reviewed by John Wilander.

Creating valid NSURLs from invalid WTF::URLs has long been a source
of issues that we find more targeted fixes for.  We do this over and
over again in different places.  This will fix all of them.

A sprinkling of little fixes is needed to keep some of the failures
the same as they were before, but this also intends to change some
of the URLs accessed after putting them into WebKit when WebKit
thinks they are not valid.

To mitigate the binary compatibility effects of this change, I did
the important changes behind a linkedOnOrAfterSDKWithBehavior check.

* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
(WTF::URL::fileSystemPath const):
(WTF::URL::emptyCFURL): Deleted.

Canonical link: https://commits.webkit.org/289946@main
